test/extensions/extension-manager.js

```diff
@@ -0,0 +1,326 @@
+        "extension-manager-unauthorized"
+      );
+
+      // Domain 2 fails
+      await checkErrorRevert(
+        extensionManager.installExtension(TEST_EXTENSION, 1, colony.address, 0, 1, 0, 2, { from: USER }),
+        "extension-manager-unauthorized"
+      );
+    });
+
+    it("does not allow an extension to be installed with a nonexistent resolver", async () => {
+      await checkErrorRevert(
+        extensionManager.installExtension(TEST_EXTENSION, 0, colony.address, 0, 1, 0, 1, { from: ROOT }),
+        "extension-manager-bad-version"
+      );
+    });
+
+    it("does not allow an extension to be installed twice", async () => {
+      await extensionManager.installExtension(TEST_EXTENSION, 1, colony.address, 0, 1, 0, 1, { from: ROOT });
+
+      await checkErrorRevert(
+        extensionManager.installExtension(TEST_EXTENSION, 1, colony.address, 0, 1, 0, 1, { from: ROOT }),
+        "extension-manager-already-installed"
+      );
+    });
+
+    it("does not allow an extension which requires root permission to be installed in a subdomain", async () => {
+      await colony.addDomain(1, 0, 1); // Domain 2
+
+      const extensionId = soliditySha3(shortid.generate());
+
+      await metaColony.addExtension(extensionManager.address, extensionId, resolver1.address, [ROOT_ROLE]);
+
+      await checkErrorRevert(
+        extensionManager.installExtension(extensionId, 1, colony.address, 0, 1, 0, 2, { from: ROOT }),
+        "extension-manager-bad-domain"
+      );
+
+      await extensionManager.installExtension(extensionId, 1, colony.address, 0, 1, 0, 1, { from: ROOT });
+    });
+
+    it("allows an extension to be installed with all roles", async () => {
+      const allRoles = [ROOT_ROLE, ARBITRATION_ROLE, ARCHITECTURE_ROLE, FUNDING_ROLE, ADMINISTRATION_ROLE];
+
+      const extensionId = soliditySha3(shortid.generate());
+
+      await metaColony.addExtension(extensionManager.address, extensionId, resolver1.address, allRoles);
+      await extensionManager.installExtension(extensionId, 1, colony.address, 0, 1, 0, 1, { from: ROOT });
+    });
+
+    it("allows an extension to be installed with non-existent roles, without erroring", async () => {
+      const fakeRole = 100;
+
+      const extensionId = soliditySha3(shortid.generate());
+
+      await metaColony.addExtension(extensionManager.address, extensionId, resolver1.address, [ROOT_ROLE, fakeRole]);
+      await extensionManager.installExtension(extensionId, 1, colony.address, 0, 1, 0, 1, { from: ROOT });
+
+      // Check that fakeRole is not part of the role bit array
+      const extensionAddress = await extensionManager.getExtension(extensionId, 1, colony.address, 1);
+      const userRoles = await colony.getUserRoles(extensionAddress, 1);
+      expect(new BN(parseInt(userRoles, 16)).and(new BN(1).shln(fakeRole))).to.be.zero;
+    });
+  });
+
+  describe("upgrading extensions", () => {
+    it("allows authorized users to upgrade an extension", async () => {
```

Comment:
I sort of tried cover all these tests in the `install` block (since the same logic does authorization for installation, upgrading, and uninstalling), but I can explicitly replicate the tests for upgrading and uninstalls as well.

Reply:
You're right, but you still have to test that the auth code is correctly wired up to the function in question, which is essentially what was missing. Otherwise you can remove the `require` that does the auth at the start of upgrade the code base still passes all tests, but is clearly broken.

Reply:
Yup, I can appreciate that distinction.